File: lib/solve_24.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cmath>
using namespace std;
class Solution{
private:
    vector<vector<string>> solutions;
    vector<string> first_solution;
    int max_generated;
public:
    vector<int> numbers;
    double target;
    Solution(vector<int> arg1){
        numbers = arg1;
        target = 24;
        max_generated = 1024;
    }
    Solution(vector<int> arg1, double arg2){
        numbers = arg1;
        target = arg2;
        max_generated = 1024;
    }
    Solution(vector<int> arg1, double arg2, int arg3){
        numbers = arg1;
        target = arg2;
        max_generated = arg3;
    }
    vector<vector<string>> get_all_solutions(){
        return solutions;
    }
    vector<string> get_first_solution(){
        return first_solution;
    }
    int get_max_generated(){
        return max_generated;
    }
    void set_max_generated(int arg1){
        max_generated = arg1;
    }
public:
    bool is_valid_input(){
        vector<double> values(numbers.begin(), numbers.end());
        return solution_exists(values, target);
    } 
    bool find_first_solution(){
        vector<double> values(numbers.begin(), numbers.end());
        vector<string> output;
        return solve_first(values, output, target);
    }
    void find_all_solutions(){
        solutions;
        vector<double> values(numbers.begin(), numbers.end());
        vector<string> output;
        solve_all(values, output, target);
        return;
    }
    void print_solutions(){
        for(int i = 0; i < solutions.size(); i++){
            cout << "Solution:" << endl;
            print_output_cpp(solutions[i]);
        }
    }
private: 
    void print_output_cpp(vector<string> output){
        for(int i = 0; i < output.size(); i++) 
            cout << output[i] << endl;
    }
private:
    bool solution_exists(vector<double>& nums, double target){
        if(nums.size() == 1){
            return fabs(nums[0] - target) < 1e-8;
        }
        double val;
        for(int i = 0; i + 1 < nums.size(); ++i){
            for(int j = i + 1; j < nums.size(); ++j){
                vector<double> new_nums;
                for(int k = 0; k < nums.size(); ++k)
                    if(k != i && k != j) new_nums.push_back(nums[k]);
                string val1 = to_string(nums[i]);
                string val2 = to_string(nums[j]); 
                //addition i, j
                val = nums[i] + nums[j];
                new_nums.push_back(val);
                if(solution_exists(new_nums, target)){
                    return true;
                }
                val = nums[i] * nums[j];
                new_nums.back() = val;
                if(solution_exists(new_nums, target)){
                    return true;
                }
                //subtraction i, j
                val = nums[i] - nums[j];
                new_nums.back() = val;
                if(solution_exists(new_nums, target)){
                    return true;
                }
                //division i, j
                val = nums[i] / nums[j];
                new_nums.back() = val;
                if(solution_exists(new_nums, target)){
                    return true;
                }
                //subtraction j, i
                val = nums[j] - nums[i];
                new_nums.back() = val;
                if(solution_exists(new_nums, target)){
                    return true;
                }
                //division j, i
                val = nums[j] / nums[i];
                new_nums.back() = val;
                if(solution_exists(new_nums, target)){
                    return true;
                }
            }
        }
        return false;
    }
    bool solve_first(vector<double>& nums, vector<string> prev_ops, double target){
        if(nums.size() == 1){
            if(fabs(nums[0] - target) < 1e-8){
                print_output_cpp(prev_ops);
                first_solution = prev_ops;
                return true;
            }
            return false;
        }
        double val;
        string input;
        for(int i = 0; i + 1 < nums.size(); ++i){
            for(int j = i + 1; j < nums.size(); ++j){
                vector<string> output;
                output = prev_ops;
                vector<double> new_nums;
                for(int k = 0; k < nums.size(); ++k)
                    if(k != i && k != j) new_nums.push_back(nums[k]);
                string val1 = to_string(nums[i]);
                string val2 = to_string(nums[j]); 
                output.push_back(val1);
                output.push_back(val2);
                //addition i, j
                val = nums[i] + nums[j];
                input = to_string(val);
                output.push_back(input);
                output.push_back("+");
                new_nums.push_back(val);
                if(solve_first(new_nums, output, target)){
                    return true;
                }
                output.pop_back();
                output.pop_back();
                val = nums[i] * nums[j];
                input = to_string(val);
                output.push_back(input);
                output.push_back("*");
                new_nums.back() = val;
                if(solve_first(new_nums, output, target)){
                    return true;
                }
                //subtraction i, j
                output.pop_back();
                output.pop_back();
                val = nums[i] - nums[j];
                input = to_string(val);
                output.push_back(input);
                output.push_back("-");
                new_nums.back() = val;
                if(solve_first(new_nums, output, target)){
                    return true;
                }
                //division i, j
                output.pop_back();
                output.pop_back();
                val = nums[i] / nums[j];
                input = to_string(val);
                output.push_back(to_string(val));
                output.push_back("/");
                new_nums.back() = val;
                if(solve_first(new_nums, output, target)){
                    return true;
                }
                output.pop_back();
                output.pop_back();
                output.pop_back();
                output.pop_back();
                output.push_back(val2);
                output.push_back(val1);
                //subtraction j, i
                val = nums[j] - nums[i];
                input = to_string(val);
                output.push_back(input);
                output.push_back("-");
                new_nums.back() = val;
                if(solve_first(new_nums, output, target)){
                    return true;
                }
                //division j, i
                output.pop_back();
                output.pop_back();
                val = nums[j] / nums[i];
                input = to_string(val);
                output.push_back(input);
                output.push_back("/");
                new_nums.back() = val;
                if(solve_first(new_nums, output, target)){
                    return true;
                }
                output.pop_back();
                output.pop_back();
                output.pop_back();
                output.pop_back();
            }
        }
        return false;
    }
    void solve_all(vector<double>& nums, vector<string> prev_ops, double target){
        if(nums.size() == 1){
            if(fabs(nums[0] - target) < 1e-8){
                //print_output_cpp(prev_ops);
                solutions.push_back(prev_ops);
            }
        }
        double val;
        string input;
        for(int i = 0; i + 1 < nums.size(); ++i){
            for(int j = i + 1; j < nums.size(); ++j){
                vector<string> output;
                output = prev_ops;
                vector<double> new_nums;
                for(int k = 0; k < nums.size(); ++k)
                    if(k != i && k != j) new_nums.push_back(nums[k]);
                string val1 = to_string(nums[i]);
                string val2 = to_string(nums[j]); 
                output.push_back(val1);
                output.push_back(val2);
                //addition i, j
                val = nums[i] + nums[j];
                input = to_string(val);
                output.push_back(input);
                output.push_back("+");
                new_nums.push_back(val);
                solve_all(new_nums, output, target);
                output.pop_back();
                output.pop_back();
                val = nums[i] * nums[j];
                input = to_string(val);
                output.push_back(input);
                output.push_back("*");
                new_nums.back() = val;
                solve_all(new_nums, output, target);
                //subtraction i, j
                output.pop_back();
                output.pop_back();
                val = nums[i] - nums[j];
                input = to_string(val);
                output.push_back(input);
                output.push_back("-");
                new_nums.back() = val;
                solve_all(new_nums, output, target);
                //division i, j
                output.pop_back();
                output.pop_back();
                val = nums[i] / nums[j];
                input = to_string(val);
                output.push_back(to_string(val));
                output.push_back("/");
                new_nums.back() = val;
                solve_all(new_nums, output, target);
                output.pop_back();
                output.pop_back();
                output.pop_back();
                output.pop_back();
                output.push_back(val2);
                output.push_back(val1);
                //subtraction j, i
                val = nums[j] - nums[i];
                input = to_string(val);
                output.push_back(input);
                output.push_back("-");
                new_nums.back() = val;
                solve_all(new_nums, output, target);
                //division j, i
                output.pop_back();
                output.pop_back();
                val = nums[j] / nums[i];
                input = to_string(val);
                output.push_back(input);
                output.push_back("/");
                new_nums.back() = val;
                solve_all(new_nums, output, target);
                output.pop_back();
                output.pop_back();
                output.pop_back();
                output.pop_back();
            }
        }
        return;
    }

};


int main(){
    //get all solutions by running find_first_solution on all combinations
    vector<int> test{5, 5, 3, 2};
    double practice_target = 24;
    Solution stupid(test, practice_target);
    //int donkey = stupid.find_first_solution() ? 1: 0;
    //stupid.find_all_solutions();
    //stupid.print_solutions();
    bool lonkey = stupid.find_first_solution();
    int donkey = lonkey ? 1 : 0;
    cout << donkey << endl;
    return 1;
    /*
    + - *
    + - *
    */
}

File: lib/solve_24.py
import math
import argparse

class Solution:
    def __init__(self, numbers, target=24, max_generated=1024):
        self.numbers = numbers
        self.target = float(target)
        self.max_generated = max_generated
        self.solutions = []
        self.first_solution = []

    def get_all_solutions(self):
        return self.solutions

    def get_first_solution(self):
        return self.first_solution

    def get_max_generated(self):
        return self.max_generated

    def set_max_generated(self, max_generated):
        self.max_generated = max_generated

    def is_valid_input(self):
        # In Python, we can directly use the numbers if they are already floats or ints.
        # The C++ version converts ints to doubles; Python handles mixed types automatically.
        return self._solution_exists([float(n) for n in self.numbers], self.target)

    def find_first_solution(self):
        self.first_solution = []
        output = []
        return self._solve_first([float(n) for n in self.numbers], output, self.target)

    def find_all_solutions(self):
        self.solutions = []
        output = []
        self._solve_all([float(n) for n in self.numbers], output, self.target)

    def print_solutions(self):
        for i, sol in enumerate(self.solutions):
            print(f"Solution {i+1}:")
            self._print_output_python(sol)

    def _print_output_python(self, output):
        for item in output:
            print(item)

    def _solution_exists(self, nums, target):
        if not nums:
            return False
        if len(nums) == 1:
            return math.isclose(nums[0], target)

        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                new_nums_base = [nums[k] for k in range(len(nums)) if k != i and k != j]
                
                # Perform operations
                # Addition
                if self._solution_exists(new_nums_base + [nums[i] + nums[j]], target):
                    return True
                # Multiplication
                if self._solution_exists(new_nums_base + [nums[i] * nums[j]], target):
                    return True
                # Subtraction (nums[i] - nums[j])
                if self._solution_exists(new_nums_base + [nums[i] - nums[j]], target):
                    return True
                # Division (nums[i] / nums[j])
                if not math.isclose(nums[j], 0):
                    if self._solution_exists(new_nums_base + [nums[i] / nums[j]], target):
                        return True
                # Subtraction (nums[j] - nums[i])
                if self._solution_exists(new_nums_base + [nums[j] - nums[i]], target):
                    return True
                # Division (nums[j] / nums[i])
                if not math.isclose(nums[i], 0):
                    if self._solution_exists(new_nums_base + [nums[j] / nums[i]], target):
                        return True
        return False

    def _solve_first(self, nums, prev_ops, target):
        if len(self.first_solution) > 0 : # an optimization from the C++ code isn't present, it's not needed as Python recursion depth is higher
            pass # Python doesn't strictly need this check as it returns immediately upon finding a solution

        if len(nums) == 1:
            if math.isclose(nums[0], target):
                self.first_solution = list(prev_ops) # Store a copy
                # self._print_output_python(prev_ops) # Assuming printing is not desired here, but in the caller
                return True
            return False

        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                new_nums_base = [nums[k] for k in range(len(nums)) if k != i and k != j]
                
                val1_str = str(nums[i])
                val2_str = str(nums[j])

                # Store original ops length to backtrack
                original_ops_len = len(prev_ops)

                # nums[i] op nums[j]
                prev_ops.extend([val1_str, val2_str])

                # Addition
                prev_ops.extend([str(nums[i] + nums[j]), "+"])
                if self._solve_first(new_nums_base + [nums[i] + nums[j]], prev_ops, target):
                    return True
                prev_ops = prev_ops[:original_ops_len+2] # Backtrack ops

                # Multiplication
                prev_ops.extend([str(nums[i] * nums[j]), "*"])
                if self._solve_first(new_nums_base + [nums[i] * nums[j]], prev_ops, target):
                    return True
                prev_ops = prev_ops[:original_ops_len+2] # Backtrack ops

                # Subtraction
                prev_ops.extend([str(nums[i] - nums[j]), "-"])
                if self._solve_first(new_nums_base + [nums[i] - nums[j]], prev_ops, target):
                    return True
                prev_ops = prev_ops[:original_ops_len+2] # Backtrack ops
                
                # Division
                if not math.isclose(nums[j], 0):
                    prev_ops.extend([str(nums[i] / nums[j]), "/"])
                    if self._solve_first(new_nums_base + [nums[i] / nums[j]], prev_ops, target):
                        return True
                    prev_ops = prev_ops[:original_ops_len+2] # Backtrack ops
                
                prev_ops = prev_ops[:original_ops_len] # Backtrack completely for swapped order

                # nums[j] op nums[i]
                prev_ops.extend([val2_str, val1_str])
                original_ops_len_swapped = len(prev_ops)


                # Subtraction
                prev_ops.extend([str(nums[j] - nums[i]), "-"])
                if self._solve_first(new_nums_base + [nums[j] - nums[i]], prev_ops, target):
                    return True
                prev_ops = prev_ops[:original_ops_len_swapped-2+2] # Backtrack ops for swapped

                # Division
                if not math.isclose(nums[i], 0):
                    prev_ops.extend([str(nums[j] / nums[i]), "/"])
                    if self._solve_first(new_nums_base + [nums[j] / nums[i]], prev_ops, target):
                        return True
                    # prev_ops = prev_ops[:original_ops_len_swapped-2+2] # Backtrack ops for swapped. Not strictly needed before full backtrack.

                prev_ops = prev_ops[:original_ops_len] # Full backtrack before next i,j pair

        return False

    def _solve_all(self, nums, prev_ops, target):
        if len(self.solutions) >= self.max_generated:
            return

        if len(nums) == 1:
            if math.isclose(nums[0], target):
                self.solutions.append(list(prev_ops)) # Store a copy
            return

        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                if len(self.solutions) >= self.max_generated:
                    return
                
                new_nums_base = [nums[k] for k in range(len(nums)) if k != i and k != j]
                
                val1_str = str(nums[i])
                val2_str = str(nums[j])
                
                original_ops_len = len(prev_ops)

                # nums[i] op nums[j]
                prev_ops.extend([val1_str, val2_str])

                # Addition
                current_op_val = nums[i] + nums[j]
                prev_ops.extend([str(current_op_val), "+"])
                self._solve_all(new_nums_base + [current_op_val], prev_ops, target)
                prev_ops = prev_ops[:original_ops_len+2] # Backtrack ops
                if len(self.solutions) >= self.max_generated: return

                # Multiplication
                current_op_val = nums[i] * nums[j]
                prev_ops.extend([str(current_op_val), "*"])
                self._solve_all(new_nums_base + [current_op_val], prev_ops, target)
                prev_ops = prev_ops[:original_ops_len+2] # Backtrack ops
                if len(self.solutions) >= self.max_generated: return

                # Subtraction
                current_op_val = nums[i] - nums[j]
                prev_ops.extend([str(current_op_val), "-"])
                self._solve_all(new_nums_base + [current_op_val], prev_ops, target)
                prev_ops = prev_ops[:original_ops_len+2] # Backtrack ops
                if len(self.solutions) >= self.max_generated: return
                
                # Division
                if not math.isclose(nums[j], 0):
                    current_op_val = nums[i] / nums[j]
                    prev_ops.extend([str(current_op_val), "/"])
                    self._solve_all(new_nums_base + [current_op_val], prev_ops, target)
                    prev_ops = prev_ops[:original_ops_len+2] # Backtrack ops
                    if len(self.solutions) >= self.max_generated: return
                
                prev_ops = prev_ops[:original_ops_len] # Backtrack completely for swapped order

                # nums[j] op nums[i]
                prev_ops.extend([val2_str, val1_str])
                # original_ops_len_swapped = len(prev_ops) # Not needed, use original_ops_len + 2

                # Subtraction
                current_op_val = nums[j] - nums[i]
                prev_ops.extend([str(current_op_val), "-"])
                self._solve_all(new_nums_base + [current_op_val], prev_ops, target)
                prev_ops = prev_ops[:original_ops_len+2] # Backtrack ops
                if len(self.solutions) >= self.max_generated: return
                
                # Division
                if not math.isclose(nums[i], 0):
                    current_op_val = nums[j] / nums[i]
                    prev_ops.extend([str(current_op_val), "/"])
                    self._solve_all(new_nums_base + [current_op_val], prev_ops, target)
                    # prev_ops = prev_ops[:original_ops_len+2] # Backtrack ops. Not strictly needed.
                    if len(self.solutions) >= self.max_generated: return

                prev_ops = prev_ops[:original_ops_len] # Full backtrack before next i,j pair
        return 

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Solve the 24 game.")
    parser.add_argument("--numbers", type=str, help="The numbers to use in the 24 game.")
    parser.add_argument("--target", type=float, default=24, help="The target value.")
    parser.add_argument("--max_generated", type=int, default=1024, help="The maximum number of solutions to generate.")
    args = parser.parse_args()

    numbers = [float(n) for n in args.numbers.split(",")]
    solution = Solution(numbers, args.target, args.max_generated)
    solution.find_all_solutions()
    solution.print_solutions()
